File: hebe-insomnia-plugin/signer.js
const aCrypto = require('crypto');

function getDigest(body) {
    if (body == null) return "";

    const hash = aCrypto.createHash('SHA256');
    hash.update(body);
    return hash.digest("base64");
}

function getSignatureValue(values, pkey) {
    const sign = aCrypto.createSign('RSA-SHA256');
    sign.update(values);
    return sign.sign("-----BEGIN PRIVATE KEY-----\n" + pkey + "\n-----END PRIVATE KEY-----", "base64");
}

function getCanonicalUrl(fullUrl) {
    const url = fullUrl.match("(api/mobile/.+)");
    if (url == null) throw 'The URL does not seem correct (does not match `(api/mobile/.+)` regex)';

    return encodeURIComponent(url[0]).toLowerCase();
}

function getHeadersList(body, digest, canonicalUrl, timestamp) {
    const signData = [
        ['vCanonicalUrl', canonicalUrl],
        body == null ? null : ['Digest', digest],
        ['vDate', new Date(timestamp + 1000).toUTCString()]
    ].filter(item => !!item);

    return {
        "headers": signData.map(item => item[0]).join(" "),
        "values": signData.map(item => item[1]).join()
    };
}

function getSignatureValues(fingerprint, pkey, body, fullUrl, timestamp) {
    const canonicalUrl = getCanonicalUrl(fullUrl);
    const digest = getDigest(body);
    const {headers, values} = getHeadersList(body, digest, canonicalUrl, timestamp);
    const signatureValue = getSignatureValue(values, pkey);

    return {
        "digest": `SHA-256=${digest}`,
        "canonicalUrl": canonicalUrl,
        "signature": `keyId="${fingerprint}",headers="${headers}",algorithm="sha256withrsa",signature=Base64(SHA256withRSA(${signatureValue}))`
    };
}

exports.getSignatureValues = getSignatureValues;

File: hebe-jvm/src/main/kotlin/signer.kt
package io.github.wulkanowy.signer.hebe

import com.migcomponents.migbase64.Base64
import java.net.URLEncoder
import java.security.KeyFactory
import java.security.Signature
import java.security.spec.PKCS8EncodedKeySpec
import java.text.SimpleDateFormat
import java.util.*
import java.security.MessageDigest.getInstance as createSign

private fun getDigest(body: String?): String {
    if (body == null) return ""
    return Base64.encodeToString(createSign("SHA-256").digest(body.toByteArray()), false)
}

private fun getSignatureValue(values: String, privateKey: String): String {
    val bl = Base64.decode(privateKey)
    val spec = PKCS8EncodedKeySpec(bl)
    val kf = KeyFactory.getInstance("RSA")

    val privateSignature = Signature.getInstance("SHA256withRSA")
    privateSignature.initSign(kf.generatePrivate(spec))
    privateSignature.update(values.toByteArray())

    return Base64.encodeToString(privateSignature.sign(), false)
}

private fun getEncodedPath(path: String): String {
    val url = ("(api/mobile/.+)".toRegex().find(path))
        ?: throw IllegalArgumentException("The URL does not seem correct (does not match `(api/mobile/.+)` regex)")

    return URLEncoder.encode(url.groupValues[0], "UTF-8").orEmpty().toLowerCase()
}

private fun getHeadersList(body: String?, digest: String, canonicalUrl: String, timestamp: Date): Pair<String, String> {
    val signData = mutableMapOf<String, String>()
    signData["vCanonicalUrl"] = canonicalUrl
    if (body != null) signData["Digest"] = digest
    signData["vDate"] = SimpleDateFormat("EEE, d MMM yyyy hh:mm:ss z", Locale.ENGLISH).apply {
        timeZone = TimeZone.getTimeZone("GMT")
    }.format(timestamp)

    return Pair(
        first = signData.keys.joinToString(" "),
        second = signData.values.joinToString("")
    )
}

fun getSignatureValues(
    fingerprint: String,
    privateKey: String,
    body: String?,
    requestPath: String,
    timestamp: Date
): Triple<String, String, String> {
    val canonicalUrl = getEncodedPath(requestPath)
    val digest = getDigest(body)
    val (headers, values) = getHeadersList(body, digest, canonicalUrl, timestamp)
    val signatureValue = getSignatureValue(values, privateKey)

    return Triple(
        "SHA-256=${digest}",
        canonicalUrl,
        """keyId="$fingerprint",headers="$headers",algorithm="sha256withrsa",signature=Base64(SHA256withRSA($signatureValue))"""
    )
}
